k2/csrc/fsa_utils_inl.h

```diff
@@ -203,26 +199,228 @@ Array1<FloatType> GetForwardScores(FsaVec &fsas, Ragged<int32_t> &state_batches,
     Ragged<int32_t> &leaving_arc_batches,
     const Array1<FloatType> *tot_scores /*= nullptr*/,
     bool log_semiring /*= true*/) {
+  K2_STATIC_ASSERT((std::is_same<float, FloatType>::value ||
+                    std::is_same<double, FloatType>::value));
+  K2_CHECK(IsCompatible(fsas, state_batches));
+  K2_CHECK(IsCompatible(fsas, leaving_arc_batches));
+  K2_CHECK_EQ(fsas.NumAxes(), 3);
+  K2_CHECK_EQ(state_batches.NumAxes(), 3);
+  K2_CHECK_EQ(leaving_arc_batches.NumAxes(), 4);
   ContextPtr &c = fsas.Context();
-  K2_LOG(INFO) << "Not Implemented!";
-  return Array1<FloatType>(c, 0);
+  int32_t num_fsas = fsas.Dim0(), num_states = fsas.TotSize(1),
+          num_arcs = fsas.TotSize(2);
+  int32_t num_batches = state_batches.Dim0();
+  // just using DCHECK below to save time in production code
+  K2_DCHECK_EQ((state_batches.TotSize(1) / num_batches), num_fsas);
+  K2_DCHECK_EQ(state_batches.NumElements(), num_states);
+  K2_DCHECK_EQ(leaving_arc_batches.Dim0(), num_batches);
+  K2_DCHECK_EQ(leaving_arc_batches.TotSize(1), state_batches.TotSize(1));
+  K2_DCHECK_EQ(leaving_arc_batches.TotSize(2), num_states);
+  K2_DCHECK_EQ(leaving_arc_batches.NumElements(), num_arcs);
+
+  FloatType negative_infinity = -std::numeric_limits<FloatType>::infinity();
+  Array1<FloatType> state_scores(c, num_states, negative_infinity);
+  FloatType *state_scores_data = state_scores.Data();
+  const int32_t *fsa_row_splits1 = fsas.RowSplits(1).Data();
+  if (tot_scores != nullptr) {
+    K2_CHECK(IsCompatible(fsas, *tot_scores));
+    K2_CHECK_EQ(tot_scores->Dim(), num_fsas);
+    const FloatType *tot_scores_data = tot_scores->Data();
+    // set the score of final state in fsa i to be negative of tot_scores[i]
+    auto lambda_set_final_state_score = [=] __host__ __device__(
+                                            int32_t fsa_idx) {
+      int32_t start_state = fsa_row_splits1[fsa_idx],
+              start_state_next_fsa = fsa_row_splits1[fsa_idx + 1];
+      if (start_state_next_fsa - start_state > 0)
+        state_scores_data[start_state_next_fsa - 1] = -tot_scores_data[fsa_idx];
+    };
+    Eval(c, num_fsas, lambda_set_final_state_score);
+  } else {
+    // set the score of final state in each fsa to be 0
+    auto lambda_set_final_state_score =
+        [=] __host__ __device__(int32_t fsa_idx) {
+          int32_t start_state = fsa_row_splits1[fsa_idx],
+                  start_state_next_fsa = fsa_row_splits1[fsa_idx + 1];
+          if (start_state_next_fsa - start_state > 0)
+            state_scores_data[start_state_next_fsa - 1] = 0;
+        };
+    Eval(c, num_fsas, lambda_set_final_state_score);
+  }
+
+  // compute leaving arc numbers in each batch
+  Array1<int32_t> leaving_arc_nums(c, num_batches);
+  Array1<int32_t> leaving_arc_start_index(c, num_batches);
+  int32_t *leaving_arc_nums_data = leaving_arc_nums.Data();
+  int32_t *leaving_arc_start_index_data = leaving_arc_start_index.Data();
+  const int32_t *fsa_in_batch_row_splits =
```

Comment:
This could would be, IMO, easier to understand if you were more rigorous about using the naming convention.  E.g. how about using names arc_batches_row_splits{1,2,3}, and batch_idx0, this_state_idx0x, next_state_idx0x, this_arc_idx0xx, next_arc_idx0xx?  I think that would make the code more understandable.

Reply:
OK, will do.

Reply:
Updated